share/github-backup-utils/ghe-backup-repositories-cluster-ng

```diff
@@ -0,0 +1,379 @@
+
++ /*/nw/??/
++ /*/nw/??/??/
++ /*/nw/??/??/??/
++ /*/nw/??/??/??/*/
++ /*/nw/??/??/??/*/*.git
++ /*/nw/??/??/??/*/*.git/refs
++ /*/nw/??/??/??/*/*.git/refs/**
++ /*/nw/??/??/??/*/*.git/logs
++ /*/nw/??/??/??/*/*.git/logs/**
+RULES
+
+    # Sync git objects and pack files. Compression is disabled during this phase
+    # since these files are already well compressed.
+  echo 1>&3
+  echo "* Transferring objects and packs ..." 1>&3
+  rsync_repository_data $1:122 $2 -H <<RULES
+- /__*__/
+- /info/
+
++ /*/
++ /*/*.git
++ /*/*.git/objects
+- /*/*.git/objects/**/tmp_*
++ /*/*.git/objects/**
+
++ /*/??/
++ /*/??/??/
++ /*/??/??/??/
++ /*/??/??/??/gist/
++ /*/??/??/??/gist/*.git
++ /*/??/??/??/gist/*.git/objects
+- /*/??/??/??/gist/*.git/objects/**/tmp_*
++ /*/??/??/??/gist/*.git/objects/**
+
++ /*/nw/??/
++ /*/nw/??/??/
++ /*/nw/??/??/??/
++ /*/nw/??/??/??/*/
++ /*/nw/??/??/??/*/*.git
++ /*/nw/??/??/??/*/*.git/objects
+- /*/nw/??/??/??/*/*.git/objects/**/tmp_*
++ /*/nw/??/??/??/*/*.git/objects/**
+RULES
+
+ echo 1>&3
+
+}
+
+# rsync all the repositories
+bm_start "$(basename $0) - Repo sync"
+for file_list in $tempdir/*.rsync; do
+  hostname=$(basename $file_list .rsync)
+
+  repo_num=$(cat $file_list | wc -l)
+  echo "* Transferring $repo_num repositories from $hostname"
+
+  sync_data $hostname $file_list &
+done
+wait
+bm_end "$(basename $0) - Repo sync"
+
+# Since there are no routes for special data directories
+# or archived repositories, we need to do this serially
+# for all hostnames. Good candidate for future optimizations.
+
+bm_start "$(basename $0) - Special Data Directories Sync"
+for h in $hostnames; do
+  # Sync __special__ data directories, including the __alambic_assets__,
+  # __hookshot__, and __purgatory__ directories. The __nodeload_archives__,
+  # __gitmon__, and __render__ directories are excludes since they act only as
+  # caches.
+  #
+  # Under v2.x and greater, only the special __purgatory__ directory remains under
+  # /data/repositories. All other special user data directories have been moved under
+  # the /data/user directory.
+  echo 1>&3
+  echo "* Transferring special data directories from $h..." 1>&3
+  rsync_repository_data $h:122 -z <<RULES
+- /__nodeload_archives__/
+- /__gitmon__/
+- /__render__/
++ /__*__/
++ /__*__/**
++ /info/
+- /info/lost+found/
++ /info/*
+RULES
+  echo 1>&3
+done
+bm_end "$(basename $0) - Special Data Directories Sync"
+
+
+bm_start "$(basename $0) - Archived Repos"
+ghe-ssh "$GHE_HOSTNAME" github-env ./bin/dgit-cluster-backup-archived-repos-routes \
+  | while read route; do
+  ar_route=$(dirname $route)
+  ghe_verbose "Got archived route $ar_route"
+  echo "$ar_route" >> $tempdir/archived_repos.rsync
+done
+
+if ! test -f $tempdir/archived_repos.rsync; then
+  echo "* No archived repositories found to backup."
+  exit 0
+fi
+
+for h in $hostnames; do
+  echo "* Transferring archived routes from $h"
+  sync_data $h $tempdir/archived_repos.rsync 2>/dev/null || true
```

Comment:
@sbryant Yes this is planned for a future update when we have some of @mutle archived repos ready. It'll use the same strategy that we use for the non-archived repos.